File: 144. Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#iterative solution
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        cur, stack = root, []
        result = []
        while cur or stack:
            if cur:
                result.append(cur.val)
                stack.append(cur.right)
                cur = cur.left
            else:
                cur = stack.pop()
        return result

#recursive solution
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        # Khởi tạo danh sách để lưu kết quả
        result = []
        
        # Hàm đệ quy để thực hiện preorder traversal
        def preorder(node):
            if not node:
                return
            # Thêm giá trị của node vào danh sách kết quả
            result.append(node.val)
            # Duyệt qua cây con trái của node
            preorder(node.left)
            # Duyệt qua cây con phải của node
            preorder(node.right)
        
        # Gọi hàm đệ quy để thực hiện preorder traversal từ root
        preorder(root)
        
        return result

File: 144. Binary Tree Preorder Traversal.c
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct Stack {
    struct TreeNode* data[10000];
    int top;
};

void push(struct Stack* stack, struct TreeNode* node) {
    stack->data[stack->top++] = node;
}

struct TreeNode* pop(struct Stack* stack) {
    return stack->data[--stack->top];
}

bool isEmpty(struct Stack* stack) {
    return stack->top == 0;
}

int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    struct Stack stack;
    stack.top = 0;
    int* result = (int*)malloc(10000 * sizeof(int));
    *returnSize = 0;
    if (root != NULL) {
        push(&stack, root);
    }
    while (!isEmpty(&stack)) {
        struct TreeNode* node = pop(&stack);
        result[(*returnSize)++] = node->val;
        if (node->right != NULL) {
            push(&stack, node->right);
        }
        if (node->left != NULL) {
            push(&stack, node->left);
        }
    }
    return result;
}

int main() {
    struct TreeNode n1, n2, n3;
    n1.val = 1; n1.left = NULL; n1.right = &n2;
    n2.val = 2; n2.left = &n3; n2.right = NULL;
    n3.val = 3; n3.left = NULL; n3.right = NULL;
    int returnSize;
    int* result = preorderTraversal(&n1, &returnSize);
    for (int i = 0; i < returnSize; i++) {
        printf("%d ", result[i]);
    }
    free(result);
    return 0;
}
